funsor/gaussian.py

```diff
@@ -165,27 +162,77 @@ def as_tensor(self):
         # Fill gaps with zeros.
         arbitrary_row = next(iter(self.parts.values()))
         prototype = next(iter(arbitrary_row.values()))
-        options = dict(dtype=prototype.dtype, device=prototype.device)
         js = set().union(*(part.keys() for part in self.parts.values()))
         rows = _find_intervals(self.parts.keys(), self.shape[-2])
         cols = _find_intervals(js, self.shape[-1])
         for i in rows:
             for j in cols:
                 if j not in self.parts[i]:
                     shape = self.shape[:-2] + (i[1] - i[0], j[1] - j[0])
-                    self.parts[i][j] = torch.zeros(shape, **options)
+                    self.parts[i][j] = ops.new_zeros(prototype, shape)
 
         # Concatenate parts.
         # TODO This could be optimized into a single .reshape().cat().reshape() if
         #   all inputs are contiguous, thereby saving a memcopy.
-        columns = {i: torch.cat([v for j, v in sorted(part.items())], dim=-1)
+        columns = {i: ops.cat([v for j, v in sorted(part.items())], dim=-1)
                    for i, part in self.parts.items()}
-        result = torch.cat([v for i, v in sorted(columns.items())], dim=-2)
+        result = ops.cat([v for i, v in sorted(columns.items())], dim=-2)
         if not torch._C._get_tracing_state():
             assert result.shape == self.shape
         return result
 
 
+def align_tensor(new_inputs, x, expand=False):
+    assert isinstance(new_inputs, OrderedDict)
+    assert isinstance(x, (Number, Tensor, Array))
+    assert all(isinstance(d.dtype, int) for d in x.inputs.values())
+
+    data = x.data
+    if isinstance(x, Number):
+        return data
+
+    old_inputs = x.inputs
+    if old_inputs == new_inputs:
+        return data
+
+    # Permute squashed input dims.
+    x_keys = tuple(old_inputs)
+    data = ops.permute(data, tuple(x_keys.index(k) for k in new_inputs if k in old_inputs) +
```

Comment:
@fritzo I just use dispatch for a hidden `_materialize` and add `permute`, `new_arange` ops so that `align_tensor` is backend agnostic.

Reply:
These look great! Do you think we should them into say a funsor/tensor_ops.py, delete the original versions in funsor/torch.py and funsor/numpy.py? That way torch.py, numpy.py and gaussian.py could all import these two generic functions.

Reply:
To me, we should do so. What is your opinion?

Reply:
Let's do it. I'm happy merging as is and deferring to a follow-up PR. Or making the change in this PR. Whichever you prefer.

Reply:
Let's just move that to a follow-up PR. This PR is already getting big 🙂

Reply:
Agreed! :)